File: frontend/src/services/api/practice.ts
import { apiClient } from './client'
import { PracticeSession, PracticeResult, PracticeType, Question } from '@/types'

const normalizeQuestion = (data: any): Question | undefined => {
  if (!data) return undefined

  return {
    questionId: data?.questionId ?? data?.question_id ?? data?.id ?? '',
    subject: data?.subject,
    topic: data?.topic,
    subtopic: data?.subtopic ?? data?.sub_topic ?? undefined,
    type: data?.type,
    difficulty: data?.difficulty,
    status: data?.status ?? 'approved',
    title: data?.title ?? '',
    description: data?.description ?? '',
    instructions: data?.instructions ?? undefined,
    options: data?.options?.map((opt: any) => ({
      id: opt?.id,
      text: opt?.text,
      isCorrect: opt?.isCorrect ?? opt?.is_correct ?? false,
    })),
    correctAnswers: data?.correctAnswers ?? data?.correct_answers ?? undefined,
    codeSnippet: data?.codeSnippet ?? data?.code_snippet ?? undefined,
    programmingLanguage: data?.programmingLanguage ?? data?.programming_language ?? undefined,
    testCases: data?.testCases ?? data?.test_cases ?? undefined,
    timeLimit: data?.timeLimit ?? data?.time_limit ?? undefined,
    memoryLimit: data?.memoryLimit ?? data?.memory_limit ?? undefined,
    expectedAnswer: data?.expectedAnswer ?? data?.expected_answer ?? undefined,
    keyPoints: data?.keyPoints ?? data?.key_points ?? undefined,
    wordLimit: data?.wordLimit ?? data?.word_limit ?? undefined,
    requirements: data?.requirements ?? undefined,
    constraints: data?.constraints ?? undefined,
    expectedComponents: data?.expectedComponents ?? data?.expected_components ?? undefined,
    diagramsRequired: data?.diagramsRequired ?? data?.diagrams_required ?? false,
    caseData: data?.caseData ?? data?.case_data ?? undefined,
    datasets: data?.datasets ?? undefined,
    analysisRequired: data?.analysisRequired ?? data?.analysis_required ?? undefined,
    explanation: data?.explanation ?? '',
    detailedExplanation: data?.detailedExplanation ?? data?.detailed_explanation ?? undefined,
    hints: data?.hints ?? [],
    commonMistakes: data?.commonMistakes ?? data?.common_mistakes ?? undefined,
    references: data?.references ?? [],
  }
}

const normalizePracticeSession = (data: any): PracticeSession => {
  return {
    sessionId: data?.sessionId ?? data?.session_id ?? data?.id ?? '',
    type: data?.type,
    status: data?.status,
    title: data?.title ?? undefined,
    description: data?.description ?? undefined,
    totalQuestions: data?.totalQuestions ?? data?.total_questions ?? 0,
    timeLimit: data?.timeLimit ?? data?.time_limit ?? data?.time_limit_minutes ?? undefined,
    passingScore: data?.passingScore ?? data?.passing_score ?? undefined,
    subject: data?.subject ?? undefined,
    topic: data?.topic ?? undefined,
    company: data?.company ?? undefined,
    role: data?.role ?? undefined,
    questionIds: data?.questionIds ?? data?.question_ids ?? [],
    currentQuestionIndex: data?.currentQuestionIndex ?? data?.current_question_index ?? 0,
    currentQuestion: normalizeQuestion(data?.currentQuestion ?? data?.current_question),
    questionsAnswered: data?.questionsAnswered ?? data?.questions_answered ?? 0,
    questionsSkipped: data?.questionsSkipped ?? data?.questions_skipped ?? 0,
    correctAnswers: data?.correctAnswers ?? data?.correct_answers ?? 0,
    incorrectAnswers: data?.incorrectAnswers ?? data?.incorrect_answers ?? 0,
    score: data?.score ?? undefined,
    accuracy: data?.accuracy ?? undefined,
    totalTimeSpent: data?.totalTimeSpent ?? data?.total_time_spent ?? undefined,
    timeElapsed: data?.timeElapsed ?? data?.time_elapsed ?? 0,
    timeRemaining: data?.timeRemaining ?? data?.time_remaining ?? undefined,
    createdAt: data?.createdAt ?? data?.created_at ?? new Date().toISOString(),
    startedAt: data?.startedAt ?? data?.started_at ?? undefined,
    completedAt: data?.completedAt ?? data?.completed_at ?? undefined,
    lastActivityAt: data?.lastActivityAt ?? data?.last_activity_at ?? undefined,
  }
}

const normalizeProgress = (data: any) => {
  return {
    current: data?.current ?? 0,
    total: data?.total ?? 0,
    correct: data?.correct ?? 0,
    incorrect: data?.incorrect ?? 0,
  }
}

const normalizeSubmitResponse = (data: any) => {
  return {
    isCorrect: data?.isCorrect ?? data?.is_correct ?? false,
    correctAnswer: data?.correctAnswer ?? data?.correct_answer ?? [],
    explanation: data?.explanation ?? '',
    nextQuestion: normalizeQuestion(data?.nextQuestion ?? data?.next_question),
    sessionCompleted: data?.sessionCompleted ?? data?.session_completed ?? false,
    progress: normalizeProgress(data?.progress),
  }
}

const normalizeSkipResponse = (data: any) => {
  return {
    nextQuestion: normalizeQuestion(data?.nextQuestion ?? data?.next_question),
    sessionCompleted: data?.sessionCompleted ?? data?.session_completed ?? false,
    progress: normalizeProgress(data?.progress),
  }
}

export const practiceService = {
  // Quick Quiz
  startQuickQuiz: async (config: {
    totalQuestions?: number
    timeLimit?: number
    subjects?: string[]
    difficulties?: string[]
    shuffle?: boolean
  }) => {
    const response = await apiClient.post<PracticeSession>('/practice/quick-quiz/start', config)
    return { ...response, data: normalizePracticeSession(response.data) }
  },

  // Topic Practice
  startTopicPractice: async (config: {
    subject: string
    topic: string
    subtopics?: string[]
    totalQuestions?: number
    timeLimit?: number
    difficulty?: string
    includeExplanations?: boolean
    shuffle?: boolean
  }) => {
    const response = await apiClient.post<PracticeSession>('/practice/topic/start', config)
    return { ...response, data: normalizePracticeSession(response.data) }
  },

  getAvailableTopics: async (subject: string) => {
    return apiClient.get(`/practice/topics/${subject}`)
  },

  // Mock Test
  startMockTest: async (config: {
    subject: string
    title?: string
    totalQuestions?: number
    timeLimit?: number
    passingScore?: number
    difficultyDistribution?: Record<string, number>
    topics?: string[]
  }) => {
    const response = await apiClient.post<PracticeSession>('/practice/mock-test/start', config)
    return { ...response, data: normalizePracticeSession(response.data) }
  },

  // Company Grid
  startCompanyPractice: async (config: {
    company: string
    role?: string
    totalQuestions?: number
    timeLimit?: number
    difficultyFocus?: string
  }) => {
    const response = await apiClient.post<PracticeSession>('/practice/company/start', config)
    return { ...response, data: normalizePracticeSession(response.data) }
  },

  getCompanyGrids: async () => {
    return apiClient.get('/practice/companies')
  },

  // Session Management
  getSession: async (sessionId: string) => {
    const response = await apiClient.get<PracticeSession>(`/practice/session/${sessionId}`)
    return { ...response, data: normalizePracticeSession(response.data) }
  },

  submitAnswer: async (sessionId: string, questionId: string, answer: any, timeTaken: number) => {
    const response = await apiClient.post(`/practice/session/${sessionId}/answer`, {
      session_id: sessionId,
      question_id: questionId,
      answer,
      time_taken_seconds: timeTaken,
    })
    return { ...response, data: normalizeSubmitResponse(response.data) }
  },

  skipQuestion: async (sessionId: string) => {
    const response = await apiClient.post(`/practice/session/${sessionId}/skip`)
    return { ...response, data: normalizeSkipResponse(response.data) }
  },

  pauseSession: async (sessionId: string) => {
    return apiClient.post(`/practice/session/${sessionId}/pause`)
  },

  resumeSession: async (sessionId: string) => {
    const response = await apiClient.post<PracticeSession>(`/practice/session/${sessionId}/resume`)
    return { ...response, data: normalizePracticeSession(response.data) }
  },

  endSession: async (sessionId: string) => {
    return apiClient.post(`/practice/session/${sessionId}/end`)
  },

  submitFeedback: async (sessionId: string, rating: number, feedback?: string) => {
    return apiClient.post(`/practice/session/${sessionId}/feedback`, {
      session_id: sessionId,
      rating,
      feedback,
    })
  },

  getSessionResults: async (sessionId: string) => {
    return apiClient.get<PracticeResult>(`/practice/session/${sessionId}/results`)
  },

  // History
  getHistory: async (days: number = 30, type?: PracticeType) => {
    let url = `/practice/history?days=${days}`
    if (type) url += `&type=${type}`
    return apiClient.get(url)
  },

  getStats: async () => {
    return apiClient.get('/practice/stats')
  },

  // Leaderboard
  getLeaderboard: async (subject: string, period: string = 'all', limit: number = 10) => {
    return apiClient.get(`/practice/leaderboard/${subject}?period=${period}&limit=${limit}`)
  },
}

File: backend/app/schemas/practice.py
"""
Practice Schemas - Pydantic models for practice-related API requests/responses
"""

from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict, Any
from datetime import datetime

from app.models.practice import PracticeType, SessionStatus

# Request Schemas
class QuickQuizRequest(BaseModel):
    """Quick quiz request"""
    total_questions: int = Field(5, ge=1, le=20)
    time_limit: int = Field(5, ge=1, le=30)
    subjects: Optional[List[str]] = None
    difficulties: Optional[List[str]] = None
    shuffle: bool = True

class TopicPracticeRequest(BaseModel):
    """Topic-wise practice request"""
    subject: str
    topic: str
    subtopics: Optional[List[str]] = None
    total_questions: int = Field(10, ge=1, le=50)
    time_limit: int = Field(10, ge=1, le=60)
    difficulty: Optional[str] = None
    include_explanations: bool = True
    shuffle: bool = True

class MockTestRequest(BaseModel):
    """Mock test request"""
    subject: str
    title: Optional[str] = None
    total_questions: int = Field(30, ge=10, le=100)
    time_limit: int = Field(30, ge=15, le=180)
    passing_score: int = Field(70, ge=0, le=100)
    difficulty_distribution: Optional[Dict[str, int]] = None
    topics: Optional[List[str]] = None

class CompanyGridRequest(BaseModel):
    """Company grid practice request"""
    company: str
    role: Optional[str] = None
    total_questions: int = Field(20, ge=5, le=50)
    time_limit: int = Field(25, ge=10, le=120)
    difficulty_focus: Optional[str] = None

class AnswerSubmission(BaseModel):
    """Answer submission for a question"""
    session_id: str
    question_id: str
    answer: Any
    time_taken_seconds: int = Field(..., ge=1, le=3600)
    hints_used: int = Field(0, ge=0)
    skipped: bool = False

class SessionAction(BaseModel):
    """Session action (pause, resume, end)"""
    session_id: str
    action: str  # pause, resume, end

class SessionFeedback(BaseModel):
    """Session feedback"""
    session_id: str
    rating: int = Field(..., ge=1, le=5)
    feedback: Optional[str] = Field(None, max_length=500)

# Response Schemas
class PracticeSessionResponse(BaseModel):
    """Practice session response"""
    session_id: str
    type: PracticeType
    status: SessionStatus
    title: Optional[str] = None
    description: Optional[str] = None
    total_questions: int
    time_limit: Optional[int] = None
    passing_score: Optional[int] = None  # 🔴 Add default None
    subject: Optional[str] = None
    topic: Optional[str] = None
    company: Optional[str] = None  # 🔴 Add default None
    
    # Current question
    current_question_index: int = 0
    current_question: Optional[Dict[str, Any]] = None
    
    # Progress
    questions_answered: int = 0
    questions_remaining: Optional[int] = None  # 🔴 Add default None
    correct_answers: int = 0
    incorrect_answers: int = 0
    time_elapsed: int = 0  # seconds
    time_remaining: Optional[int] = None  # seconds
    
    # Timestamps
    started_at: Optional[datetime] = None
    created_at: datetime
    
    # 🔴 Add validator to ensure all fields have values
    @validator('questions_remaining', always=True)
    def set_questions_remaining(cls, v, values):
        if v is None and 'total_questions' in values and 'questions_answered' in values:
            return values['total_questions'] - values['questions_answered']
        return v
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }
        # 🔴 Allow extra fields from backend
        extra = "allow"

class PracticeSessionDetailResponse(PracticeSessionResponse):
    """Detailed practice session response"""
    questions: List[Dict[str, Any]] = []
    question_status: Dict[str, Dict[str, Any]] = {}
    time_per_question: List[int] = []
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }
        extra = "allow"

class PracticeResultResponse(BaseModel):
    """Practice session results response"""
    session_id: str
    type: PracticeType
    score: float = 0.0
    accuracy: float = 0.0
    total_time_spent: int = 0
    correct_answers: int = 0
    incorrect_answers: int = 0
    skipped_questions: int = 0
    
    # Detailed performance
    subject_wise: Dict[str, Dict[str, Any]] = {}
    topic_wise: Dict[str, Dict[str, Any]] = {}
    difficulty_wise: Dict[str, Dict[str, Any]] = {}
    
    # Analysis
    strengths: List[str] = []
    weaknesses: List[str] = []
    recommendations: List[str] = []
    
    # Comparison
    percentile: Optional[float] = None
    average_score: Optional[float] = None
    
    completed_at: datetime
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }
        extra = "allow"

class PracticeHistoryResponse(BaseModel):
    """Practice history response"""
    total_sessions: int = 0
    total_time_spent: int = 0  # minutes
    average_score: float = 0.0
    average_accuracy: float = 0.0
    recent_sessions: List[Dict[str, Any]] = []
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }
        extra = "allow"

class CompanyGridInfo(BaseModel):
    """Company grid information"""
    company: str
    description: Optional[str] = None
    total_questions: int = 0
    topics: List[str] = []
    difficulty_breakdown: Dict[str, int] = {}
    average_company_score: float = 0.0
    user_score: Optional[float] = None
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }
        extra = "allow"